layabout.py

```diff
@@ -1,3 +1,287 @@
+from inspect import signature
+from functools import wraps
+from collections import defaultdict
+
+from slackclient import SlackClient
+
+# TODO: This is a dependency of slackclient needed for exception handling. In
+# the future, we might be able to remove this. For now we don't want to declare
+# it as an explicit dependency.
+from websocket import WebSocketConnectionClosedException
+
 __author__ = 'Reilly Tucker Siemens'
 __email__ = 'reilly@tuckersiemens.com'
-__version__ = '0.1.0'
+__version__ = '1.0.0b'
+
+# Private type alias for the complex type of the handlers defaultdict.
+_Handlers = DefaultDict[str, List[Tuple[Callable, dict]]]
+
+log = logging.getLogger(__name__)
+
+
+class LayaboutError(Exception):
+    """ Base error for all Layabout exceptions. """
+
+
+class MissingSlackToken(LayaboutError):
+    """ Raised if a Slack API token could not be found. """
+
+
+class FailedConnection(LayaboutError, ConnectionError):
+    """
+    Raised if the Slack client could not connect to the Slack API.
+
+    Inherits from both :obj:`LayaboutError` and the built-in
+    :obj:`ConnectionError` for convenience in exception handling.
+    """
+
+
+class Layabout:
+    """
+    An event handler on top of the Slack RTM API.
+
+    Args:
+        name: A unique name for this :obj:`Layabout` instance.
+        env_var: The environment variable to try to load a Slack API token
+            from.
+
+    Attributes:
+        name (str): A unique name for this :obj:`Layabout` instance.
+        slack (SlackClient): A :obj:`slackclient.SlackClient` instance.
+            Initially :obj:`None`, the attribute is set after calling the
+            :obj:`Layabout.connect` or :obj:`Layabout.run`
+            methods.
+
+    Example:
+
+        .. code-block:: python
+
+           from layabout import Layabout
+
+           layabout = Layabout('app')
+
+
+           @layabout.handle('message')
+           def echo(slack, event):
+               \"\"\" Echo all messages seen by the app. \"\"\"
+               channel = event['channel']
+               message = event['text']
+               subtype = event.get('subtype')
+
+               # Avoid an infinite loop of echoing our own messages.
+               if subtype != 'bot_message':
+                   slack.rtm_send_message(channel, message)
+
+           layabout.run()
+    """
+    def __init__(self, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
+        # TODO: Consider keyword only arguments.
+        self.name = name
+        self._env_var = env_var
+        self._token: str = None
+        self.slack: SlackClient = None
+        self._handlers: _Handlers = defaultdict(list)
+
+    def handle(self, type: str, kwargs: dict = None) -> Callable:
+        """
+        Register an event handler with the :obj:`Layabout` instance.
+
+        Args:
+            type: The name of a Slack RTM API event to be handled. As a
+                special case, although it is not a proper RTM event, ``*`` may
```

Comment:
Is there a defined order in which events will be handled or is it purely first come first serve?